#ifndef NFT_H
#define NFT_H

#include <string>
#include <vector>
#include <cstdint>
#include "generated/nft.pb.h"

class NFT {
public:
    // ğŸ†” Core NFT fields
    std::string id;
    std::string creator;
    std::string owner;
    std::string metadata;
    std::string imageHash;
    int64_t timestamp;
    std::vector<uint8_t> signature;

    // ğŸ” zk-STARK proof
    std::vector<uint8_t> zkStarkProof;

    // ğŸ§  Optional fields
    std::string version;
    std::string nft_type;
    std::string proof_hash;
    std::string extra_data;
    std::vector<std::string> previous_versions;
    // ğŸ‘¤ Creator identity (e.g., name, alias)
    std::string creator_identity;

    // ğŸ“¦ Bundled assets (e.g., image filenames, asset IDs)
    std::vector<std::string> bundledAssets;

    // ğŸ“œ Transfer ledger
    std::vector<std::string> transferHistory;

    // ğŸ” Dual-signature (Dilithium)
    std::vector<uint8_t> dilithium_signature;

    // â³ Expiry and Revocation
    int64_t expiry_timestamp = 0;
    bool revoked = false;

    // ğŸ”’ Encrypted metadata (optional)
    std::string encrypted_metadata;

    NFT() = default;

    // ğŸ” Serialization
    NFTProto toProto() const;
    bool fromProto(const NFTProto& proto);
    std::string toJSON() const;
    static NFT fromJSON(const std::string& jsonStr);

    // âœ… Verification
    bool verifySignature() const;
    bool verifyZkStarkProof() const;

    // âš™ï¸ Proof generation
    void generateZkStarkProof();

    // ğŸ“¤ Metadata submission
    bool submitMetadataHashTransaction() const;

    // ğŸ“¦ Export
    bool exportToFile(const std::string& filename = "") const;

   std::string getSignatureMessage() const {
    return id + creator + owner + metadata + imageHash + std::to_string(timestamp);
  }

};

// === Standalone utility declarations for reMint ===
std::string calculateHash(const std::string& input);
std::string generateZkStarkProof(const std::string& metadata, const std::string& imageHash, const std::string& creator);
bool submitMetadataHashTransaction(const std::string& metadataHash, const std::string& creator,
                                   const std::string& signatureScheme, bool isReMint);
void exportNFTtoFile(const std::string& filename, const std::string& metadataHash,
                     const std::string& creator, const std::string& version,
                     const std::string& zkProof);
bool reMintNFT(const std::string& creator,
               const std::string& prevNftId,
               const std::string& newMetadata,
               const std::string& imageHash,
               const std::string& signatureScheme,
               const std::string& previousVersion,
               const std::string& previousZkProof);

#endif // NFT_H
